# **LIMPIEZA DE DATOS**



## Configuración inicial

In [1]:
# from google.colab import drive
# drive.mount('/content/drive')

In [2]:
import pandas as pd
from pathlib import Path
import numpy as np

## Carga de base Parquet



In [3]:
# pip install pyarrow

In [4]:
# Cargamos base Parquet y revisamos tipos de variables
PARQUET_PATH = r"C:\Users\Hugo Jazyel\OneDrive\INEI\PEU-CD\Cursos especializados\Proyecto Integrador\enaho_merged.parquet"

df = pd.read_parquet(PARQUET_PATH)

print("✅ Archivo Parquet cargado correctamente.")
print("Filas:", df.shape[0], "| Columnas:", df.shape[1])
print("\nTipos de datos (muestra):")
print(df.dtypes)

✅ Archivo Parquet cargado correctamente.
Filas: 438428 | Columnas: 38

Tipos de datos (muestra):
anio             object
conglome         object
vivienda         object
hogar            object
codperso         object
txt505           object
txt505b          object
txt506           object
p208a              int8
p501            float64
p502            float64
p503            float64
p504            float64
p5041           float64
p5042           float64
p5043           float64
p5044           float64
p5045           float64
p5046           float64
p5047           float64
p5048           float64
p5049           float64
p50410          float64
p50411          float64
p207               int8
p301a           float64
p506r4          float64
p507            float64
p510            float64
p505r4          float64
p207_label       object
p301a_label      object
p506r4_label     object
p507_label       object
p510_label       object
p505r4_label     object
p301a1          float64
p301a1_label   

In [5]:
# Filtramos solo personas con ocupación, ya que solo estos son considerados en la p505
cols_filter = [
    "p501", "p502", "p503", "p5041", "p5042", "p5043", "p5044",
    "p5045", "p5046", "p5047", "p5048", "p5049", "p50410", "p50411"
]

# Aseguramos que existan en la base
cols_exist = [c for c in cols_filter if c in df.columns]

# Filtro: 1 en cualquiera de las columnas y sin NA
mask_valid = df[cols_exist].apply(lambda x: (x == 1) & (~x.isna())).any(axis=1)

# Aplicamos filtro
df = df.loc[mask_valid].copy()

print(f"✅ Filtro aplicado: {df.shape[0]} filas restantes tras excluir NA y mantener == 1.")
df.shape

✅ Filtro aplicado: 316025 filas restantes tras excluir NA y mantener == 1.


(316025, 38)

In [6]:
df.shape

(316025, 38)

In [7]:
print(df.columns.to_list())

['anio', 'conglome', 'vivienda', 'hogar', 'codperso', 'txt505', 'txt505b', 'txt506', 'p208a', 'p501', 'p502', 'p503', 'p504', 'p5041', 'p5042', 'p5043', 'p5044', 'p5045', 'p5046', 'p5047', 'p5048', 'p5049', 'p50410', 'p50411', 'p207', 'p301a', 'p506r4', 'p507', 'p510', 'p505r4', 'p207_label', 'p301a_label', 'p506r4_label', 'p507_label', 'p510_label', 'p505r4_label', 'p301a1', 'p301a1_label']


## Conversión de variables numéricas (coercivo y seguro)

In [8]:
# ================================================
# Convertimos variables de interes
# ================================================
# Variables de identificación (IDs) -> Las convertimos a category para ahorrar memoria
df['anio'] = df['anio'].astype('category')
df['conglome'] = df['conglome'].astype('category')
df['vivienda'] = df['vivienda'].astype('category')
df['hogar'] = df['hogar'].astype('category')
df['codperso'] = df['codperso'].astype('category')

# Variables de texto (ocupación) -> Las mantenemos como string
df.loc[df['txt505'].str.strip() == '', 'txt505'] = np.nan
df.loc[df['txt505b'].str.strip() == '', 'txt505b'] = np.nan
df.loc[df['txt506'].str.strip() == '', 'txt506'] = np.nan
df['txt505'] = df['txt505'].astype('string')
df['txt505b'] = df['txt505b'].astype('string')
df['txt506'] = df['txt506'].astype('string')

# Convertimos a string y rellenamos con ceros a la izquierda. Luego a category
df['p505r4'] = df['p505r4'].fillna(0).astype('int64').astype('str').str.zfill(4)
df['p506r4'] = df['p506r4'].fillna(0).astype('int64').astype('str').str.zfill(4)

df['p505r4'] = df['p505r4'].astype('category')
df['p506r4'] = df['p506r4'].astype('category')

# Convertimos a category las de pocas categorias
df['p301a'] = df['p301a'].astype('category')   # 12 categorías
df['p301a1'] = df['p301a1'].astype('category')
df['p507'] = df['p507'].astype('category')      # 7 categorías
df['p510'] = df['p510'].astype('category')      # 6 categorías

# Convertimos a string los labels
df['p207_label'] = df['p207_label'].astype('string')
df['p301a_label'] = df['p301a_label'].astype('string')
df['p505r4_label'] = df['p505r4_label'].astype('string')
df['p506r4_label'] = df['p506r4_label'].astype('string')
df['p507_label'] = df['p507_label'].astype('string')
df['p510_label'] = df['p510_label'].astype('string')
df['p301a1_label'] = df['p301a1_label'].astype('string')

print("Tipos tras conversión:")
print(df.dtypes)

Tipos tras conversión:
anio                  category
conglome              category
vivienda              category
hogar                 category
codperso              category
txt505          string[python]
txt505b         string[python]
txt506          string[python]
p208a                     int8
p501                   float64
p502                   float64
p503                   float64
p504                   float64
p5041                  float64
p5042                  float64
p5043                  float64
p5044                  float64
p5045                  float64
p5046                  float64
p5047                  float64
p5048                  float64
p5049                  float64
p50410                 float64
p50411                 float64
p207                      int8
p301a                 category
p506r4                category
p507                  category
p510                  category
p505r4                category
p207_label      string[python]
p301a_label     

In [9]:
print(df.columns.to_list())

['anio', 'conglome', 'vivienda', 'hogar', 'codperso', 'txt505', 'txt505b', 'txt506', 'p208a', 'p501', 'p502', 'p503', 'p504', 'p5041', 'p5042', 'p5043', 'p5044', 'p5045', 'p5046', 'p5047', 'p5048', 'p5049', 'p50410', 'p50411', 'p207', 'p301a', 'p506r4', 'p507', 'p510', 'p505r4', 'p207_label', 'p301a_label', 'p506r4_label', 'p507_label', 'p510_label', 'p505r4_label', 'p301a1', 'p301a1_label']


## Separando variables de interés

In [10]:
# Separando las variables de interes
cols_of_interest = ['anio', 'conglome', 'vivienda', 'hogar', 'codperso', 'p207', 'p207_label', 'p208a', 'p301a', 'p301a_label', 'p301a1', 'p301a1_label', 'p505r4', 'p505r4_label', 'txt505', 'txt505b', 'p506r4', 'p506r4_label', 'txt506', 'p507', 'p507_label', 'p510', 'p510_label']
df = df[cols_of_interest]

In [11]:
# ===========================================================
# Verificación de columnas seleccionadas y orden final
# ===========================================================
print(f"📊 El DataFrame actual tiene {df.shape[1]} columnas:")

for i, col in enumerate(df.columns, start=1):
    print(f"{i:>2}. {col}")

📊 El DataFrame actual tiene 23 columnas:
 1. anio
 2. conglome
 3. vivienda
 4. hogar
 5. codperso
 6. p207
 7. p207_label
 8. p208a
 9. p301a
10. p301a_label
11. p301a1
12. p301a1_label
13. p505r4
14. p505r4_label
15. txt505
16. txt505b
17. p506r4
18. p506r4_label
19. txt506
20. p507
21. p507_label
22. p510
23. p510_label


In [12]:
df.shape

(316025, 23)

In [13]:
print(df.columns.to_list(),"\n")
print(df.dtypes)

['anio', 'conglome', 'vivienda', 'hogar', 'codperso', 'p207', 'p207_label', 'p208a', 'p301a', 'p301a_label', 'p301a1', 'p301a1_label', 'p505r4', 'p505r4_label', 'txt505', 'txt505b', 'p506r4', 'p506r4_label', 'txt506', 'p507', 'p507_label', 'p510', 'p510_label'] 

anio                  category
conglome              category
vivienda              category
hogar                 category
codperso              category
p207                      int8
p207_label      string[python]
p208a                     int8
p301a                 category
p301a_label     string[python]
p301a1                category
p301a1_label    string[python]
p505r4                category
p505r4_label    string[python]
txt505          string[python]
txt505b         string[python]
p506r4                category
p506r4_label    string[python]
txt506          string[python]
p507                  category
p507_label      string[python]
p510                  category
p510_label      string[python]
dtype: object


In [14]:
df.tail(5)

anio conglome vivienda hogar codperso  p207 p207_label  p208a p301a  \
438422  2024   021001      026    11       03     1     hombre     17   6.0   
438423  2024   021001      026    11       04     2      mujer     15   5.0   
438424  2024   021001      029    11       01     1     hombre     80   3.0   
438425  2024   021001      037    11       01     1     hombre     56   5.0   
438426  2024   021001      037    11       02     2      mujer     54   3.0   

                  p301a_label  ...  \
438422    secundaria completa  ...   
438423  secundaria incompleta  ...   
438424    primaria incompleta  ...   
438425  secundaria incompleta  ...   
438426    primaria incompleta  ...   

                                             p505r4_label  \
438422  vendedores minoristas en tiendas y establecimi...   
438423     peones de explotaciones agrícolas y ganaderas   
438424  agricultores y trabajadores calificados de cul...   
438425  agricultores y trabajadores calificados de cul...   
438426     peones de explotaciones agrícolas y ganaderas   

                                    txt505  \
438422        AYUDANTE DE VENDEDOR DE TELA   
438423               AYUDANTE DE ÑRODUCTOR   
438424                           PRODUCTOR   
438425              PRODUCTOR AGROPECUARIO   
438426  AYUDANTE DE PRODUCTOR AGROPECUARIO   

                                                  txt505b p506r4  \
438422                                    ATENDER CLIENTE   4751   
438423                                       CISECHAR PAP   0113   
438424                                       AGROPECUARIO   0113   
438425                                       APORTAR PAPA   0150   
438426  ALIMENTAR PORCINO AMARRAR VACUNO OVINO ORDEÑAR...   0150   

                                             p506r4_label  \
438422  venta al por menor de productos textiles en al...   
438423  cultivo de vegetales y melones, raíces y tubÃ...   
438424  cultivo de vegetales y melones, raíces y tubÃ...   
438425  cultivo de productos agrícolas en combinació...   
438426  cultivo de productos agrícolas en combinació...   

                                                   txt506 p507  \
438422                VENTA DE TELA AL POR MENOR ENTIENDA  3.0   
438423                                    CULTIVO DE PAPA  5.0   
438424                                    CULTIVO DE PAPA  2.0   
438425  CULTIVA PAPA HABA AVENA FORRAJE IZAÑO CRIANZA ...  2.0   
438426  CULTIVP DE PAPA AVENA FORRAJERA ALFA CRANZA DE...  5.0   

                               p507_label p510                 p510_label  
438422                           empleado  6.0  empresa o patrono privado  
438423  trabajador familiar no remunerado  NaN                       <NA>  
438424           trabajador independiente  NaN                       <NA>  
438425           trabajador independiente  NaN                       <NA>  
438426  trabajador familiar no remunerado  NaN                       <NA>  

[5 rows x 23 columns]

In [15]:
print("Tamaño real:", df.shape)
print("Primeros índices:", df.index[:5].tolist())
print("Últimos índices:", df.index[-5:].tolist())
print("Índice único y ordenado:", df.index.is_monotonic_increasing and df.index.is_unique)

Tamaño real: (316025, 23)
Primeros índices: [0, 2, 3, 6, 7]
Últimos índices: [438422, 438423, 438424, 438425, 438426]
Índice único y ordenado: True


In [16]:
df = df.reset_index(drop=True)

In [17]:
df.tail()

anio conglome vivienda hogar codperso  p207 p207_label  p208a p301a  \
316020  2024   021001      026    11       03     1     hombre     17   6.0   
316021  2024   021001      026    11       04     2      mujer     15   5.0   
316022  2024   021001      029    11       01     1     hombre     80   3.0   
316023  2024   021001      037    11       01     1     hombre     56   5.0   
316024  2024   021001      037    11       02     2      mujer     54   3.0   

                  p301a_label  ...  \
316020    secundaria completa  ...   
316021  secundaria incompleta  ...   
316022    primaria incompleta  ...   
316023  secundaria incompleta  ...   
316024    primaria incompleta  ...   

                                             p505r4_label  \
316020  vendedores minoristas en tiendas y establecimi...   
316021     peones de explotaciones agrícolas y ganaderas   
316022  agricultores y trabajadores calificados de cul...   
316023  agricultores y trabajadores calificados de cul...   
316024     peones de explotaciones agrícolas y ganaderas   

                                    txt505  \
316020        AYUDANTE DE VENDEDOR DE TELA   
316021               AYUDANTE DE ÑRODUCTOR   
316022                           PRODUCTOR   
316023              PRODUCTOR AGROPECUARIO   
316024  AYUDANTE DE PRODUCTOR AGROPECUARIO   

                                                  txt505b p506r4  \
316020                                    ATENDER CLIENTE   4751   
316021                                       CISECHAR PAP   0113   
316022                                       AGROPECUARIO   0113   
316023                                       APORTAR PAPA   0150   
316024  ALIMENTAR PORCINO AMARRAR VACUNO OVINO ORDEÑAR...   0150   

                                             p506r4_label  \
316020  venta al por menor de productos textiles en al...   
316021  cultivo de vegetales y melones, raíces y tubÃ...   
316022  cultivo de vegetales y melones, raíces y tubÃ...   
316023  cultivo de productos agrícolas en combinació...   
316024  cultivo de productos agrícolas en combinació...   

                                                   txt506 p507  \
316020                VENTA DE TELA AL POR MENOR ENTIENDA  3.0   
316021                                    CULTIVO DE PAPA  5.0   
316022                                    CULTIVO DE PAPA  2.0   
316023  CULTIVA PAPA HABA AVENA FORRAJE IZAÑO CRIANZA ...  2.0   
316024  CULTIVP DE PAPA AVENA FORRAJERA ALFA CRANZA DE...  5.0   

                               p507_label p510                 p510_label  
316020                           empleado  6.0  empresa o patrono privado  
316021  trabajador familiar no remunerado  NaN                       <NA>  
316022           trabajador independiente  NaN                       <NA>  
316023           trabajador independiente  NaN                       <NA>  
316024  trabajador familiar no remunerado  NaN                       <NA>  

[5 rows x 23 columns]

## Validaciones rápidas de integridad

In [18]:
# ================================================
# Validaciones de integridad y rangos
# ================================================
# --- Resumen tabular de columnas str ---
def summarize_string(df):
    rows = []
    str_cols = df.select_dtypes(include=["object", "string"]).columns

    for c in str_cols:
        s = df[c].astype("string")

        lens = s.dropna().str.len()

        row = {
            "col": c,
            "dtype": df[c].dtype.name,
            "n": len(s),
            "n_missing": int(s.isna().sum()),
            "%missing": 100 * s.isna().mean(),
            "n_unique": s.nunique(dropna=True),
            "len_min": lens.min(skipna=True),
            "len_p25": lens.quantile(0.25),
            "len_p50": lens.quantile(0.50),
            "len_p75": lens.quantile(0.75),
            "len_max": lens.max(skipna=True),
        }
        rows.append(row)

    out = pd.DataFrame(rows)
    out = out.sort_values("%missing", ascending=False).reset_index(drop=True)
    return out


str_summary = summarize_string(df)

str_summary.style.format({
    "%missing": "{:.2f}",
    "len_min": "{:.0f}", "len_p25": "{:.0f}",
    "len_p50": "{:.0f}", "len_p75": "{:.0f}", "len_max": "{:.0f}"
})

In [19]:
# --- Resumen tabular de columnas numéricas ---

def summarize_numeric(df):
    rows = []
    num_cols = df.select_dtypes(include=["number"]).columns

    for c in num_cols:
        s = df[c]

        # Convertir categorías a numérico (códigos)
        if isinstance(s.dtype, pd.CategoricalDtype):
            s = s.cat.codes.replace(-1, np.nan)

        s = pd.to_numeric(s, errors="coerce")

        row = {
            "col": c,
            "dtype": df[c].dtype.name,
            "n": len(s),
            "n_missing": int(s.isna().sum()),
            "%missing": 100 * s.isna().mean(),
            "n_unique": s.nunique(dropna=True),
            "min": s.min(skipna=True),
            "p01": s.quantile(0.01),
            "p25": s.quantile(0.25),
            "p50": s.quantile(0.50),
            "p75": s.quantile(0.75),
            "p99": s.quantile(0.99),
            "max": s.max(skipna=True),
            "mean": s.mean(skipna=True),
            "std": s.std(skipna=True),
        }
        rows.append(row)

    out = pd.DataFrame(rows)
    out = out.sort_values("%missing", ascending=False).reset_index(drop=True)
    return out


num_summary = summarize_numeric(df)

num_summary.style.format({
    "%missing": "{:.2f}",
    "min": "{:.0f}", "p01": "{:.0f}", "p25": "{:.0f}", "p50": "{:.0f}",
    "p75": "{:.0f}", "p99": "{:.0f}", "max": "{:.0f}",
    "mean": "{:.2f}", "std": "{:.2f}"
})

In [20]:
# --- Resumen tabular de columnas category ---

def summarize_category(df, top_k=5):
    """
    Resume columnas de tipo 'category'.
    Para cada columna devuelve:
      - n, n_missing, %missing, n_unique
      - is_ordered
      - n_codes (número de categorías definidas)
      - top_k_labels (string "label1 (pct%), label2 (pct%), ...")
      - top_k_counts (string "count1, count2, ...")
    """
    rows = []
    cat_cols = df.select_dtypes(include=["category"]).columns

    for c in cat_cols:
        s = df[c].astype("category")
        n = len(s)
        n_missing = int(s.isna().sum())
        pct_missing = 100 * s.isna().mean()
        n_unique = s.nunique(dropna=True)
        is_ordered = bool(s.cat.ordered)
        n_codes = len(s.cat.categories)

        # top_k con porcentajes sobre el total no-missing (o sobre n si prefieres)
        vc = s.dropna().value_counts()
        top = vc.head(top_k)
        top_labels = []
        top_counts = []
        for label, cnt in top.items():
            pct = 100 * cnt / (n - n_missing) if (n - n_missing) > 0 else 0.0
            top_labels.append(f"{label} ({pct:.1f}%)")
            top_counts.append(str(int(cnt)))

        row = {
            "col": c,
            "dtype": df[c].dtype.name,
            "n": n,
            "n_missing": n_missing,
            "%missing": pct_missing,
            "n_unique": n_unique,
            "is_ordered": is_ordered,
            "n_codes": n_codes,
            f"top_{top_k}_labels": ", ".join(top_labels) if top_labels else "",
            f"top_{top_k}_counts": ", ".join(top_counts) if top_counts else ""
        }
        rows.append(row)

    out = pd.DataFrame(rows)
    out = out.sort_values("%missing", ascending=False).reset_index(drop=True)
    return out

# Uso:
cat_summary = summarize_category(df, top_k=5)

# Mostrar con formato
cat_summary.style.format({
    "%missing": "{:.2f}"
})

In [21]:
# ===========================================================
# ✅ Verificación de reglas de negocio (versión corregida y robusta)
# ===========================================================
checks = []

# --- Reglas numéricas ---
if "anio" in df:
    s = pd.to_numeric(df["anio"], errors="coerce")  # 🔹 Conversión segura a numérico
    ok = s.between(2020, 2024).all()
    checks.append({
        "check": "anio in [2020,2024]",
        "status": "OK" if ok else "ALERTA",
        "detalle": f"min={s.min()}, max={s.max()}, n_missing={s.isna().sum()}"
    })

if "p208a" in df:
    s = pd.to_numeric(df["p208a"], errors="coerce")
    ok = s.dropna().between(10, 99).all()
    checks.append({
        "check": "p208a in [10,99]",
        "status": "OK" if ok else "ALERTA",
        "detalle": f"min={int(s.min())}, p01={int(s.quantile(0.01))}, p99={int(s.quantile(0.99))}, max={int(s.max())}, n_missing={s.isna().sum()}"
    })

for c in ["p505r4", "p506r4"]:
    if c in df:
        s = pd.to_numeric(df[c], errors="coerce")
        ok = (s.dropna() > 0).all()
        checks.append({
            "check": f"{c} > 0",
            "status": "OK" if ok else "ALERTA",
            "detalle": f"min={int(s.min())}, max={int(s.max())}, n_missing={int(s.isna().sum())}"
        })

if "p207" in df:
    s = pd.to_numeric(df["p207"], errors="coerce")
    vals = sorted(s.dropna().unique().tolist())
    ok = set(vals).issubset({1, 2})
    checks.append({
        "check": "p207 in {1,2}",
        "status": "OK" if ok else "ALERTA",
        "detalle": f"valores={vals}, n_missing={s.isna().sum()}"
    })

# --- Reglas categóricas ---
for col in df.select_dtypes("category"):
    n_unique = df[col].nunique(dropna=True)
    top_val = df[col].value_counts(dropna=True).index[0] if n_unique > 0 else None
    top_freq = df[col].value_counts(dropna=True).iloc[0] if n_unique > 0 else 0
    checks.append({
        "check": f"{col} categoría",
        "status": "OK" if n_unique > 0 else "ALERTA",
        "detalle": f"n_unique={n_unique}, top='{top_val}', top_freq={top_freq}, n_missing={df[col].isna().sum()}"
    })

# --- Reglas de texto ---
for col in df.select_dtypes("object"):
    s = df[col].astype(str)
    lens = s.str.len()
    checks.append({
        "check": f"{col} longitud texto",
        "status": "OK" if lens.dropna().mean() > 0 else "ALERTA",
        "detalle": f"len_min={lens.min()}, len_p25={lens.quantile(0.25):.1f}, len_p50={lens.median():.1f}, len_p75={lens.quantile(0.75):.1f}, len_max={lens.max()}, n_missing={df[col].isna().sum()}"
    })

# --- Resultado final ---
checks_df = pd.DataFrame(checks)
display(checks_df)


check status  \
0   anio in [2020,2024]     OK   
1      p208a in [10,99]     OK   
2            p505r4 > 0     OK   
3            p506r4 > 0     OK   
4         p207 in {1,2}     OK   
5        anio categoría     OK   
6    conglome categoría     OK   
7    vivienda categoría     OK   
8       hogar categoría     OK   
9    codperso categoría     OK   
10      p301a categoría     OK   
11     p301a1 categoría     OK   
12     p505r4 categoría     OK   
13     p506r4 categoría     OK   
14       p507 categoría     OK   
15       p510 categoría     OK   

                                              detalle  
0                     min=2020, max=2024, n_missing=0  
1         min=14, p01=14, p99=82, max=98, n_missing=0  
2                      min=111, max=9629, n_missing=0  
3                      min=111, max=9900, n_missing=0  
4                         valores=[1, 2], n_missing=0  
5   n_unique=5, top='2022', top_freq=65240, n_miss...  
6   n_unique=9189, top='016630', top_freq=133, n_m...  
7   n_unique=766, top='010', top_freq=2986, n_miss...  
8   n_unique=15, top='11', top_freq=310611, n_miss...  
9   n_unique=18, top='01', top_freq=141672, n_miss...  
10  n_unique=12, top='6.0', top_freq=81622, n_miss...  
11  n_unique=802, top='714035.0', top_freq=5255, n...  
12  n_unique=450, top='9211', top_freq=65703, n_mi...  
13  n_unique=392, top='0150', top_freq=88074, n_mi...  
14  n_unique=7, top='2.0', top_freq=123078, n_miss...  
15  n_unique=6, top='6.0', top_freq=90200, n_missi...

## Tratamiento de missings

In [22]:
((df.isna().sum()*100) / df.shape[0]).loc[lambda x : x > 0].sort_values(ascending=False).round(2)

p301a1          73.98
p301a1_label    73.98
p510            62.36
p510_label      62.36
txt505b          0.14
p301a_label      0.02
p301a            0.02
dtype: float64

### Variables `301a_label` y `301a1_label`

- `301a_label`: Último año o grado de estudios y nivel que aprobó -> 0.02% de missings

- `301a1_label`: Carrera superior universitaria o no universitaria que ud. estudia o ha estudiado -> 73.98% de missings

In [23]:
# Observar presencia exclusiva de NAs

df[df['p301a1_label'].isna()][['p301a_label', 'p301a1_label']].drop_duplicates()

p301a_label p301a1_label
0                        primaria completa         <NA>
2                      secundaria completa         <NA>
7     superior no universitaria incompleta         <NA>
9       superior no universitaria completa         <NA>
14                     primaria incompleta         <NA>
22                               sin nivel         <NA>
38         superior universitaria completa         <NA>
49                   secundaria incompleta         <NA>
61       superior universitaria incompleta         <NA>
73                      maestria/doctorado         <NA>
1345                               inicial         <NA>
4042                      básica especial         <NA>
8048                                  <NA>         <NA>

In [24]:
# Lista de valores que quieres filtrar
valores_filtrados = [
    'superior no universitaria incompleta',
    'superior no universitaria completa',
    'superior universitaria incompleta',
    'superior universitaria completa'
]

# Filtrar el DataFrame con esas categorías
df_filtrado = df[df['p301a_label'].isin(valores_filtrados)]

# Calcular % de missings por cada p301a_label
porcentaje_missings = (
    df_filtrado['p301a1_label'].isna()
    .groupby(df_filtrado['p301a_label'])
    .mean()
    .reset_index(name='porc_missing_p301a1')
)

# Convertir a porcentaje
porcentaje_missings['porc_missing_p301a1'] = (
    porcentaje_missings['porc_missing_p301a1'] * 100
).round(2)

# Mostrar resultado
print(porcentaje_missings)


                            p301a_label  porc_missing_p301a1
0    superior no universitaria completa                10.57
1  superior no universitaria incompleta                10.30
2       superior universitaria completa                 9.37
3     superior universitaria incompleta                 8.42


Es decir, en realidad tenemos un 38.66% de missings para la variable `301a1_label` (Carrera superior universitaria o no universitaria que ud. estudia o ha estudiado).

### Variable `p510_label`

- `p510_label`: Tipo de entidad para la que trabajó en su ocupación principal.

El encuestador llena esta pregunta si en la pregunta `p507_label` (Ud. se desempeñó en su ocupación principal como) marcó "Empleado", "Obrero" u "Otro".

In [25]:
# Observar presencia exclusiva de NAs

df[df['p510_label'].isna()][['p507_label', 'p510_label']].drop_duplicates()

p507_label p510_label
0            trabajador independiente       <NA>
2                 empleador o patrono       <NA>
21  trabajador familiar no remunerado       <NA>
23               trabajador del hogar       <NA>

In [26]:
df['p510_label'].isna().groupby(df['p507_label']).value_counts()


p507_label                         p510_label
empleado                           False          57552
empleador o patrono                True           10072
obrero                             False          59889
otro                               False           1514
trabajador del hogar               True            4768
trabajador familiar no remunerado  True           59152
trabajador independiente           True          123078
Name: count, dtype: int64

In [27]:
tabla = (
    df.assign(missing=df['p510_label'].isna())
    .groupby('p507_label')['missing']
    .agg(total='count', n_missing='sum')
    .reset_index()
)

tabla['porc_missing_p510'] = (tabla['n_missing'] / tabla['total'] * 100).round(2)

print(tabla)


                          p507_label   total  n_missing  porc_missing_p510
0                           empleado   57552          0                0.0
1                empleador o patrono   10072      10072              100.0
2                             obrero   59889          0                0.0
3                               otro    1514          0                0.0
4               trabajador del hogar    4768       4768              100.0
5  trabajador familiar no remunerado   59152      59152              100.0
6           trabajador independiente  123078     123078              100.0


Validamos que sí tenemos todos los valores (100%) para la `p510_label` cuando en la pregunta `p507_label` marcaron "Empleado", "Obrero" u "Otro".

### Variable `txt505b`

- `txt505b`: Tareas que realizó en su ocupación principal.

In [28]:
df[df['txt505b'].isna()][['txt505b', 'txt505']].drop_duplicates()

txt505b                          txt505
85        <NA>                PEON  DE ALBAÑIL
1642      <NA>          CONDUCTOR DE VEHICULO 
1656      <NA>                        COCINERO
1744      <NA>                   RECEPCIONISTA
1864      <NA>                 PEON DE CULTIVO
...        ...                             ...
60340     <NA>            PROFESOR DE PRIMARIA
108787    <NA>  SUB OFICIAL TÉCNICO DE PRIMERA
113594    <NA>          PRODUCTOR AGROPECUARIO
223959    <NA>           TRABAJADORA DEL HOGAR
232147    <NA>              PRODUCTOR AGRICOLA

[98 rows x 2 columns]

Tenemos un 0.14% de missings en la `txt505b`, lo cual puede reducir el contexto para las distintas 98 ocupaciones.

## Limpieza de data



In [29]:
# Distribución de longitudes de literales en base a caracteres.
longitudes = df['txt505'].astype(str).str.len()
longitudes.describe()

count    316025.000000
mean         23.784786
std          11.830971
min           1.000000
25%          17.000000
50%          21.000000
75%          30.000000
max          70.000000
Name: txt505, dtype: float64

In [30]:
# Literales que contienen numeros
print(df['txt505'][df['txt505'].str.contains(r'\d+', na=False)].shape[0])
df['txt505'][df['txt505'].str.contains(r'\d+', na=False)]

12


30823                                     PE0N AGROPECUARIO
47403                                 CONDUCT0R DE TRICICLO
70778                   ELABORADOR DE MAQUETA VIRTUAL EN 3D
104483                                   SUB OFICIAL DE 3RA
119099                                                    5
120887                                                    4
125011    ASESOR DE ESTUDIENTES PARA MANEJO DE PROGRAMAS...
161763                                  AUXILIAR DE 0FICINA
183231                                   SUB OFICIAL DE 2DA
208595                                        ASES0RA LEGAL
231030                                CHOFER  DE CAMIO0NETA
311454                VENDEDORA DE CERVEZA GASEOSA AGUA MI5
Name: txt505, dtype: string

In [31]:
# Literales que presentan solo numeros
df[~df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False)]

anio conglome vivienda hogar codperso  p207 p207_label  p208a p301a  \
119099  2021   019731      024    11       03     2      mujer     15   5.0   
120887  2021   019955      039    11       02     2      mujer     38   1.0   

                  p301a_label  ...  \
119099  secundaria incompleta  ...   
120887              sin nivel  ...   

                                          p505r4_label txt505  \
119099              panaderos, pasteleros y confiteros      5   
120887  vendedores en kiosco fijo y puestos de mercado      4   

                              txt505b p506r4  \
119099         PREPARAR DECORAR TORTA   1071   
120887  VENDER OFRECER COBRAR ATENDER   4782   

                                             p506r4_label  \
119099            elaboración de productos de panadería   
120887  venta al por menor de productos textiles, pren...   

                                                   txt506 p507  \
119099   PREPARACION VENTA DE TORTAS EN CASA DEL CLIENTES  2.0   
120887  VENTA DE ZAPATILLA ZAPATO AL POR MENOR EN PUES...  3.0   

                      p507_label p510                 p510_label  
119099  trabajador independiente  NaN                       <NA>  
120887                  empleado  6.0  empresa o patrono privado  

[2 rows x 23 columns]

In [32]:
# Eliminacion de Literales que presentan solo numeros
df = df[df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False)].copy()

In [33]:
# Literales con 2 caracteres o menos
df[
    df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False) &
    (df['txt505'].str.len() <= 2)
][['txt505', 'p505r4','txt505b','p506r4_label']]

txt505 p505r4                                            txt505b  \
16748      DJ   3521                  ARMAR EQUIPO PONER MEZCLAR MÚSICA   
38119      DJ   3521                      HACER MEZCLA DE MUSICA ANIMAR   
153615     DJ   3521  PROGRAMAR MÚSICA EN UNA EN UNA FIESTA SOCIAL C...   
176176     DJ   3521         CLASIFICAR DISCO PONER MUSICA EN DISCOTECA   
183372     DJ   3521                        MESCLAR MUSICA PONER MUSICA   
195123     DJ   3521             AYUDAR ACOMODAR EQUIPOS MEZCLAR MUSICA   
216573     DJ   3521                PROGRAMAR LA MUSICA DE LA DISCOTECA   
223569     DJ   3521                          REALIZAR MEZCLA DE MUSICA   
231940     DJ   3521                                        PONER MUSIC   
233531     DJ   3521                      PONER MUSICA EQUALIZAR SONIDA   
234283     DJ   3521                                       PONER MUSICA   
276370     DJ   5169                      SELECCIONAR Y PONER LA MUSICA   
276439     DJ   3521                                       PONER MUSICA   
278748     DJ   5169          PONER PROGRAMAR MUSICA ENTRETENER PUBLICO   
278991     DJ   3521                       MEZCLAR MUSICA ANIMAR FIESTA   
296410     DJ   3521                  REALIZAR VENDER PISTAS DE MUSICAS   
306968     DG   3521              INSTALAR CONECTAR EQUIPO PONER MÚSICA   
309164     DJ   3521  SELECCIONAR MEZCLAR REPRODUCIR MUSICA EN EVENT...   

                                             p506r4_label  
16748   actividades de arte, entretenimiento y creativ...  
38119   actividades de arte, entretenimiento y creativ...  
153615  actividades de arte, entretenimiento y creativ...  
176176  otras actividades de diversión y esparcimient...  
183372  otras actividades de diversión y esparcimient...  
195123  otras actividades de diversión y esparcimient...  
216573  otras actividades de diversión y esparcimient...  
223569  actividades de arte, entretenimiento y creativ...  
231940  otras actividades de diversión y esparcimient...  
233531  actividades de arte, entretenimiento y creativ...  
234283  otras actividades de diversión y esparcimient...  
276370  actividades de arte, entretenimiento y creativ...  
276439  restaurantes y servicios móviles de alimentac...  
278748  actividades de arte, entretenimiento y creativ...  
278991  actividades de arte, entretenimiento y creativ...  
296410  actividades de arte, entretenimiento y creativ...  
306968  actividades de arte, entretenimiento y creativ...  
309164  actividades de arte, entretenimiento y creativ...

In [34]:
# Literales con 3 caracteres o menos que no incluya los anteriores revisados
df[
    df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False) &
    (df['txt505'].str.len() <= 3) &
    (~df['txt505'].isin(['DJ', 'DG']))
][['txt505', 'txt505b', 'p506r4_label']]

txt505                                         txt505b  \
244015    SUP  PAGAR PERSONAL PROVEEDORES ADMINISTRAR NEGOCIO   

                                             p506r4_label  
244015  actividades de arquitectura e ingeniería y ac...

In [35]:
#Eliminacion de registro
df = df[df['txt505'].str.upper() != 'SUP'].copy()

In [36]:
# Literales con 3 caracteres o menos que no incluya los anteriores revisados
df[
    df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False) &
    (df['txt505'].str.len() <= 4) &
    (~df['txt505'].isin(['DJ', 'DG']))
][['txt505', 'txt505b', 'p506r4_label']]

txt505                                            txt505b  \
4307     CHEF          PREPARAR INGREDIENTES PELAR PICAR COCINAR   
4487     MOZA                                      ATENDER MESAS   
4710     MOZO              ATENDER OFRECER CLIENTE LLEVAR PLATOS   
4969     MOZO                     ATENDER CLIENTES TOMAR PEDIDOS   
4989     PEON                       LIMPIAR CANALES SACAR BASURA   
...       ...                                                ...   
313388   MOZO  LLEVAR PLATOS PASAR COPAS LAVAR PLATOS PREPARA...   
314748   PEON                             LIMPIAR CANAL COSECHAR   
314756   PEON                   AYUDANTE DE CHANCADORA DE PIEDRA   
315404   PEON                              ECHANDO FERTILIZANTES   
315930   MOZO                             ATENDER LIMPIAR COBRAR   

                                             p506r4_label  
4307    restaurantes y servicios móviles de alimentac...  
4487    otras actividades de diversión y esparcimient...  
4710    restaurantes y servicios móviles de alimentac...  
4969    restaurantes y servicios móviles de alimentac...  
4989                  actividades de apoyo a los cultivos  
...                                                   ...  
313388                   provisión de comidas preparadas  
314748  cultivo de vegetales y melones, raíces y tubÃ...  
314756          construcción de caminos y vías férreas  
315404  cultivo de vegetales y melones, raíces y tubÃ...  
315930  restaurantes y servicios móviles de alimentac...  

[775 rows x 3 columns]

In [37]:
# Literales con 4 caracteres o menos que no incluya los anteriores revisados
df[
    df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False) &
    (df['txt505'].str.len() <= 4) &
    (~df['txt505'].isin(['DJ', 'DG', 'SUP']))
][['txt505', 'txt505b','p506r4_label']].drop_duplicates(subset='txt505').reset_index(drop=True)

txt505                                          txt505b  \
0    CHEF        PREPARAR INGREDIENTES PELAR PICAR COCINAR   
1    MOZA                                    ATENDER MESAS   
2    MOZO            ATENDER OFRECER CLIENTE LLEVAR PLATOS   
3    PEON                     LIMPIAR CANALES SACAR BASURA   
4    TAXI                                         CONDUCIR   
5    MOSA  ATENDER CLIENTE PREPARAR REFRESCO LIMPIAR LOCAL   
6    JUEZ                   LLEVAR ACABO JUCIOS AUDIENCIAS   
7    BUZO                          EXTRAER PESCADO DEL MAR   
8    JEFE                       GESTIONAR TRAMITAR DIRIGIR   
9    MOSO                    MOSEAR TOMAR ENTREGAR PEDIDOS   
10   PEÓN                    DESHIERBAR PARCELA DE BOLAINA   
11   CAJA                    RECEPCIONAR DINER PAGO DE LUZ   
12   NANA                             CUIDAR  DAR DE COMER   
13   MOZQ                               ATENDER AL PUBLICO   

                                         p506r4_label  
0   restaurantes y servicios móviles de alimentac...  
1   otras actividades de diversión y esparcimient...  
2   restaurantes y servicios móviles de alimentac...  
3                 actividades de apoyo a los cultivos  
4    otros tipos de transporte terrestre de pasajeros  
5   restaurantes y servicios móviles de alimentac...  
6   actividades de mantenimiento del orden públic...  
7                                     pesca marítima  
8   actividades de la administración pública en ...  
9   restaurantes y servicios móviles de alimentac...  
10                            otros cultivos perennes  
11  producción, transmisión y distribución de e...  
12  actividades de los hogares en calidad de emple...  
13  restaurantes y servicios móviles de alimentac...

In [38]:
#CORREGIR CASOS OBVIOS
correcciones = {
    'DG': 'DJ',
    'MOSO': 'MOZO',
    'MOSA': 'MOZA',
    'MOZQ': 'MOZA',   # caso detectado
    'PEON': 'PEÓN',
    'CAJA': 'CAJERA',  # aquí depende del contexto, si quieres lo conversamos
}
df['txt505'] = df['txt505'].replace(correcciones)

In [39]:
# Literales con 4 caracteres o menos que no incluya los anteriores revisados
df[
    df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False) &
    (df['txt505'].str.len() <= 4) &
    (~df['txt505'].isin(['DJ', 'DG', 'SUP']))
][['txt505', 'txt505b','p506r4_label']].drop_duplicates(subset='txt505').reset_index(drop=True)

txt505                                    txt505b  \
0   CHEF  PREPARAR INGREDIENTES PELAR PICAR COCINAR   
1   MOZA                              ATENDER MESAS   
2   MOZO      ATENDER OFRECER CLIENTE LLEVAR PLATOS   
3   PEÓN               LIMPIAR CANALES SACAR BASURA   
4   TAXI                                   CONDUCIR   
5   JUEZ             LLEVAR ACABO JUCIOS AUDIENCIAS   
6   BUZO                    EXTRAER PESCADO DEL MAR   
7   JEFE                 GESTIONAR TRAMITAR DIRIGIR   
8   NANA                       CUIDAR  DAR DE COMER   

                                        p506r4_label  
0  restaurantes y servicios móviles de alimentac...  
1  otras actividades de diversión y esparcimient...  
2  restaurantes y servicios móviles de alimentac...  
3                actividades de apoyo a los cultivos  
4   otros tipos de transporte terrestre de pasajeros  
5  actividades de mantenimiento del orden públic...  
6                                    pesca marítima  
7  actividades de la administración pública en ...  
8  actividades de los hogares en calidad de emple...

In [40]:
# Literales con 5 caracteres o menos que no incluya los anteriores revisados
df[
    df['txt505'].str.contains(r'[a-zA-ZñÑáéíóúÁÉÍÓÚ]', na=False) &
    (df['txt505'].str.len() <= 5) &
    (~df['txt505'].isin(['DJ', 'DG', 'SUP']))
][['txt505', 'txt505b','p506r4_label']].drop_duplicates(subset='txt505').reset_index(drop=True)

txt505                                            txt505b  \
0    CHEF          PREPARAR INGREDIENTES PELAR PICAR COCINAR   
1    MOZA                                      ATENDER MESAS   
2    MOZO              ATENDER OFRECER CLIENTE LLEVAR PLATOS   
3    PEÓN                       LIMPIAR CANALES SACAR BASURA   
4    TAXI                                           CONDUCIR   
5   CHEFF                                    PREPARAR COMIDA   
6   VIGIA  CONTROLAR PASE DE INGRESO Y SALIDA DE LOS VEHI...   
7    JUEZ                     LLEVAR ACABO JUCIOS AUDIENCIAS   
8    BUZO                            EXTRAER PESCADO DEL MAR   
9   DIJEY  PROGRAMAR MUSICA ANIMAR MUSICA EN REUNIONES PA...   
10   JEFE                         GESTIONAR TRAMITAR DIRIGIR   
11  PINTO                                             PINTAR   
12  DILER                REPARTIR CARTAS A JUGADORES DE MESA   
13   NANA                               CUIDAR  DAR DE COMER   
14  DIJEI  REALIZAR PLANEACION ORGANIZACION DE EVENTO MUS...   
15  TUTOR                               DESARROLLO DE EXAMEN   
16  DIYEI              CONTROLAR MEZCLAR MUSICA EN DISCOTECA   
17  ACTOR                                   ACTUAR COORDINAR   
18  PEONA                                           COSECHAR   
19  MOZZA  AYUDAR A SERVIR COMIDA COBRAR Y DAR VUELTO A C...   
20  MESEO                                  ATENDER A CLIENTE   

                                         p506r4_label  
0   restaurantes y servicios móviles de alimentac...  
1   otras actividades de diversión y esparcimient...  
2   restaurantes y servicios móviles de alimentac...  
3                 actividades de apoyo a los cultivos  
4    otros tipos de transporte terrestre de pasajeros  
5   restaurantes y servicios móviles de alimentac...  
6           construcción de caminos y vías férreas  
7   actividades de mantenimiento del orden públic...  
8                                     pesca marítima  
9   actividades de arte, entretenimiento y creativ...  
10  actividades de la administración pública en ...  
11                terminación y acabado de edificios  
12                    actividades de juego y apuestas  
13  actividades de los hogares en calidad de emple...  
14  actividades de arte, entretenimiento y creativ...  
15                   otros tipos de enseñanza n.c.p.  
16  otras actividades de diversión y esparcimient...  
17  actividades de arte, entretenimiento y creativ...  
18                                    cultivo de uvas  
19  restaurantes y servicios móviles de alimentac...  
20  restaurantes y servicios móviles de alimentac...

## Limpieza lingüística

**Objetivo**: corregir de manera programática, masiva y lingüísticamente robusta las tres columnas de texto (`txt505`, `txt505b`, `txt506`):

- Errores de codificación (como “comunicación” → “comunicación” y “niño” → “niño”).
→ Esto se debe a lecturas UTF-8 interpretadas como Latin-1 (mojibake).

- Palabras que deben llevar tilde obligatoriamente (como comunicación, educación, situación…).
→ Solo esas, no los casos donde el acento es opcional (si/sí, solo/sólo).

- Ñ incorrectas o perdidas (ej. nino → niño, espana → españa).

- Espacios múltiples, caracteres invisibles o de control.

In [41]:
pip install pandas ftfy pyspellchecker wordfreq unidecode python-Levenshtein swifter tqdm pyarrow fastparquet

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [42]:
# ===========================================================
# BLOQUE FINAL – Limpieza y corrección lingüística avanzada (versión estable combinada)
# ===========================================================
!pip install ftfy pyspellchecker wordfreq unidecode python-Levenshtein swifter > /dev/null

import pandas as pd, re, unicodedata, ftfy, Levenshtein, swifter
from spellchecker import SpellChecker
from wordfreq import zipf_frequency, top_n_list
from unidecode import unidecode

# -----------------------------------------------------------
# 0️⃣ Configuración base
# -----------------------------------------------------------
df_sub = df.copy()   # ⚙️ Cambia a df para correr toda la base

spell = SpellChecker(language="es")
spanish_vocab = set(top_n_list("es", 100_000))

# -----------------------------------------------------------
# 1️⃣ Restaurar “ñ” (versión robusta)
# -----------------------------------------------------------
def restaurar_n(text):
    words, corrected = text.split(), []
    for w in words:
        if not w.isalpha():
            corrected.append(w)
            continue
        if "ñ" in w:
            corrected.append(w)
            continue

        # Probamos reemplazar letras "n" por "ñ"
        variants = {w[:i]+"ñ"+w[i+1:] for i,ch in enumerate(w) if ch=="n"}
        valid = [v for v in variants if (zipf_frequency(v, "es") > 1.8 and v in spanish_vocab)]
        if valid and zipf_frequency(valid[0], "es") > zipf_frequency(w, "es"):
            corrected.append(valid[0])
        else:
            corrected.append(w)
    return " ".join(corrected)

# -----------------------------------------------------------
# 2️⃣ Restaurar tildes (versión híbrida, evita verbos conjugados)
# -----------------------------------------------------------
def restaurar_tildes(text):
    words, corrected = text.split(), []
    for w in words:
        if not w.isalpha():
            corrected.append(w)
            continue

        plain = unidecode(w)
        if plain == w:
            best_variant, best_freq = None, 0
            for base, acc in zip("aeiou", "áéíóú"):
                if base in w:
                    variant = w.replace(base, acc)
                    f = zipf_frequency(variant, "es")
                    if f > best_freq:
                        best_variant, best_freq = variant, f

            if best_variant:
                termina = w[-3:]
                termina_vocal = w[-1] in "aeiou"
                es_verbo = termina in ["o", "as", "os", "amos", "emos", "imos"]

                # --- Evita acentuar verbos o sustantivos terminados en vocal
                if not es_verbo and (not termina_vocal) and best_freq - zipf_frequency(w, "es") > 0.8:
                    corrected.append(best_variant)
                else:
                    corrected.append(w)
            else:
                corrected.append(w)
        else:
            corrected.append(w)
    return " ".join(corrected)

# -----------------------------------------------------------
# 3️⃣ Corrección ortográfica segura
# -----------------------------------------------------------
def corregir_seguro(token):
    if not token.isalpha():
        return token
    if len(token) <= 3:
        return token
    if token in spanish_vocab:
        return token

    suggestion = spell.correction(token)
    if suggestion is None:
        return token

    dist = Levenshtein.distance(token, suggestion)
    freq_sug = zipf_frequency(suggestion, "es")
    freq_tok = zipf_frequency(token, "es")

    plain = unidecode(token)
    for base, acc in zip("aeiou", "áéíóú"):
        if base in token:
            variant = token.replace(base, acc)
            if zipf_frequency(variant, "es") > zipf_frequency(token, "es"):
                return variant

    if dist <= 1 and freq_sug > freq_tok + 0.5:
        return suggestion
    return token

# -----------------------------------------------------------
# 4️⃣ Limpieza y corrección integral
# -----------------------------------------------------------
def limpiar_y_corregir(text):
    if not isinstance(text, str):
        return text
    text = ftfy.fix_text(text)
    text = unicodedata.normalize("NFC", text)
    text = re.sub(r"\s+", " ", text.lower()).strip()

    tokens = [corregir_seguro(t) for t in text.split()]
    text = " ".join(tokens)
    text = restaurar_n(text)
    text = restaurar_tildes(text)
    text = re.sub(r"\s{2,}", " ", text).strip()
    return text

# -----------------------------------------------------------
# 5️⃣ Aplicación paralelizada
# -----------------------------------------------------------
for c in ["txt505", "txt505b", "txt506"]:
    if c in df_sub.columns:
        print(f"⚙️ Corrigiendo columna: {c} ...")
        df_sub[f"{c}_clean"] = df_sub[c].fillna("").astype(str).swifter.apply(limpiar_y_corregir)

# -----------------------------------------------------------
# 6️⃣ Unificación final
# -----------------------------------------------------------
df_sub["texto_final"] = (
    df_sub["txt505_clean"].fillna("") + " " +
    df_sub["txt505b_clean"].fillna("") + " " +
    df_sub["txt506_clean"].fillna("")
).str.replace(r"\s{2,}", " ", regex=True).str.strip().replace("", pd.NA)

# -----------------------------------------------------------
# 7️⃣ Vista de resultados
# -----------------------------------------------------------
pd.set_option("display.max_colwidth", 120)
print("\n✅ Limpieza y corrección lingüística avanzada completada (versión híbrida).")
display(df_sub[["txt505","txt505b","txt506","texto_final"]].sample(10, random_state=42))


El sistema no puede encontrar la ruta especificada.
c:\Users\Hugo Jazyel\AppData\Local\Programs\Python\Python313\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


⚙️ Corrigiendo columna: txt505 ...


Pandas Apply: 100%|██████████| 316022/316022 [1:34:17<00:00, 55.86it/s]  


⚙️ Corrigiendo columna: txt505b ...


Pandas Apply: 100%|██████████| 316022/316022 [5:00:44<00:00, 17.51it/s]    


⚙️ Corrigiendo columna: txt506 ...


Pandas Apply: 100%|██████████| 316022/316022 [13:04:06<00:00,  6.72it/s]      



✅ Limpieza y corrección lingüística avanzada completada (versión híbrida).


txt505  \
104908                             PROMOTOR DE VENTAS   
3892                            AYUDANTE AGROPECUARIO   
158463                         PRODUCTOR AGROPECUARIO   
130769                              AYUDANTE PECUARIO   
87795                           AYUDANTE AGROPECUARIA   
105899                     OPERARIO DE METAL MECANICA   
23688     AYUDANTE DE PREPARADORA VENDEDORA DE COMIDA   
40554   AYUDANTE DE VENDEDOR DE GAS PETRÓLEO GASOLINA   
218262                                  PEON AGRICOLA   
277073                           PERSONAL DE LIMPIEZA   

                                                              txt505b  \
104908                              VENDER OFRECER  PLANES DE EQUIPOS   
3892             PASTEAR JUNTAR AMARRAR OVINO PORCINO DESHIERBAR MAIZ   
158463                   CORTAR MALEZA ECHAR ABONO ALIMENTAR ANIMALES   
130769                                  AYUDAR ATENDER AVES DE CORRAL   
87795   DESHIERBAR CULTIVOS DE MAIZ DAR AGUA ALIMENTOS A LOS ANIMALES   
105899                                          ARMAR PIEZAS DE METAL   
23688                 CORTAR PICAR LIMPIAR MEZA DOBLAR LAS SERVILLETA   
40554                           AYUDAR A VENDER GAS PETRÓLEO GASOLINA   
218262                                                  CARGAR SANDIA   
277073                                                  LIMPIAR BAÑOS   

                                                                        txt506  \
104908                VENTA POR MENOR DE EQUIPOS TELEFONICOS Y PLANES EN LOCAL   
3892                      CULTIVO MAIZ PAPA TRIGO CRIANZA EQUINO OVINO PORCINO   
158463       CULTIVO DE MAIZ PAPA CRIANZA DE CUY AVES DE CORRAL GANADO PORCINO   
130769                 CRIANZA DE GANADO OVINO AUQUENIDO VACUNO AVES DE CORRAL   
87795   CULTIVO MAIZ FRIJOL YUCA CACAO CRIANZA DE VACUNO EQUINO PORCINO AVES D   
105899                                                TALLER DE METAL MECANICA   
23688                                PREPARACION VENTA DE COMIDA EN RESTAURANT   
40554               VENTA DE GAS PETRÓLEO GASOLINA AL POR MAYOR MENOR EN GRIFO   
218262                                                       CULTIVO DE SANDIA   
277073                           CULTIVO Y EXPORTACIÓN DE UVA ARÁNDANOS FRESCO   

                                                                                                                    texto_final  
104908             promotor de ventas vender ofrecer planes de equipos venta por menor de equipos telefonicos y planes en local  
3892    ayudante agropecuario pasear juntar amarrar ovino porcino deshierbar maíz cultivo maíz papa trigo crianza equino ovi...  
158463  productor agropecuario cortar maleza echar abono alimentar animales cultivo de maíz papa crianza de cuy aves de corr...  
130769                  ayudante pecuario ayudar atender aves de corral crianza de ganado ovino auquénido vacuno aves de corral  
87795   ayudante agropecuaria deshierbar cultivos de maíz dar agua alimentos a los animales cultivo maíz frijol yuca cacao c...  
105899                                                operario de metal mecanica armar piezas de metal taller de metal mecanica  
23688   ayudante de preparadora vendedora de comida cortar picar limpiar meza doblar las servilleta preparación venta de com...  
40554   ayudante de vendedor de gas petróleo gasolina ayudar a vender gas petróleo gasolina venta de gas petróleo gasolina a...  
218262                                                                            peón agricola cargar sandia cultivo de sandia  
277073                                         personal de limpieza limpiar baños cultivo y exportación de uva arándanos fresco

In [43]:
display(df_sub[["p505r4_label","p506r4_label"]].sample(10, random_state=42))

p505r4_label  \
104908                                                                       representantes comerciales   
3892                                                     peones de explotaciones agrícolas y ganaderas   
158463  agricultores y trabajadores calificados de cultivos mixtos (agrícolas, pecuarios y forestales)   
130769                                                   peones de explotaciones agrícolas y ganaderas   
87795                                                    peones de explotaciones agrícolas y ganaderas   
105899                                                                            herreros y forjadores   
23688                                                                               ayudantes de cocina   
40554                                                                             vendedores mayoristas   
218262                                                   peones de explotaciones agrícolas y ganaderas   
277073                           limpiadores y asistentes de oficinas, hoteles y otros establecimientos   

                                                                                                                   p506r4_label  
104908  venta al por menor de pc, unidades periféricas, prog informáticos y eq de telecomunicaciones en almacenes especial...  
3892                              cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)  
158463                            cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)  
130769                                                                                                cría de  ovejas y cabras  
87795                             cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)  
105899                                                                fabricación de productos metálicos para uso estructural  
23688                                                                        restaurantes y servicios móviles de alimentación  
40554                                     venta al por mayor de combustibles sólidos, líquidos y gaseosos y productos conexos  
218262                                                                    cultivo de vegetales y melones, raíces y tubérculos  
277073                                                                                                          cultivo de uvas

In [44]:
import ftfy

df_sub["p505r4_label"] = df_sub["p505r4_label"].astype(str).apply(ftfy.fix_text)
display(df_sub[["p505r4_label"]].sample(5, random_state=42))

p505r4_label
104908                                                                      representantes comerciales
3892                                                     peones de explotaciones agrícolas y ganaderas
158463  agricultores y trabajadores calificados de cultivos mixtos (agrícolas, pecuarios y forestales)
130769                                                   peones de explotaciones agrícolas y ganaderas
87795                                                    peones de explotaciones agrícolas y ganaderas

In [45]:
df_sub["p506r4_label"] = df_sub["p506r4_label"].astype(str).apply(ftfy.fix_text)
display(df_sub[["p506r4_label"]].sample(5, random_state=42))

p506r4_label
104908  venta al por menor de pc, unidades periféricas, prog informáticos y eq de telecomunicaciones en almacenes especializ...
3892                                  cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)
158463                                cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)
130769                                                                                                 cría de  ovejas y cabras
87795                                 cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)

In [46]:
display(df_sub[["p505r4_label","p506r4_label"]].sample(10, random_state=42))

p505r4_label  \
104908                                                                      representantes comerciales   
3892                                                     peones de explotaciones agrícolas y ganaderas   
158463  agricultores y trabajadores calificados de cultivos mixtos (agrícolas, pecuarios y forestales)   
130769                                                   peones de explotaciones agrícolas y ganaderas   
87795                                                    peones de explotaciones agrícolas y ganaderas   
105899                                                                           herreros y forjadores   
23688                                                                              ayudantes de cocina   
40554                                                                            vendedores mayoristas   
218262                                                   peones de explotaciones agrícolas y ganaderas   
277073                          limpiadores y asistentes de oficinas, hoteles y otros establecimientos   

                                                                                                                   p506r4_label  
104908  venta al por menor de pc, unidades periféricas, prog informáticos y eq de telecomunicaciones en almacenes especializ...  
3892                                  cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)  
158463                                cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)  
130769                                                                                                 cría de  ovejas y cabras  
87795                                 cultivo de productos agrícolas en combinación con la cría de animales (explotación mixta)  
105899                                                                  fabricación de productos metálicos para uso estructural  
23688                                                                          restaurantes y servicios móviles de alimentación  
40554                                       venta al por mayor de combustibles sólidos, líquidos y gaseosos y productos conexos  
218262                                                                      cultivo de vegetales y melones, raíces y tubérculos  
277073                                                                                                          cultivo de uvas

In [49]:
display(df_sub[["p301a1_label","p510_label", "p301a_label", "p507_label"]].sample(50, random_state=42))

p301a1_label  \
104908                                    <NA>   
3892                                      <NA>   
158463                                    <NA>   
130769                                    <NA>   
87795                                     <NA>   
105899                     mecanica automotriz   
23688                     ingeniería ambiental   
40554                                     <NA>   
218262                                    <NA>   
277073                                    <NA>   
92733                                     <NA>   
281676                                    <NA>   
81081                                     <NA>   
175664                                    <NA>   
196868                                    <NA>   
17525                                     <NA>   
154354                                    <NA>   
183663    administración y ciencias policiales   
203010                electrotecnia industrial   
169875                                    <NA>   
14569                                     <NA>   
302628                                    <NA>   
19139                                     <NA>   
310140                                    <NA>   
180787                                    <NA>   
120391                                    <NA>   
83616                                     <NA>   
247373                        ingeniería civil   
46570                                     <NA>   
256909                  ingeniería estadistica   
217143                                    <NA>   
276711                                    <NA>   
190574                                    <NA>   
102254                                    <NA>   
186593                        ingeniería civil   
249194                                    <NA>   
285429                                    <NA>   
144479                                    <NA>   
88030                      laboratorio clinico   
99914                                     <NA>   
99331                                     <NA>   
276315               mecanica de mantenimiento   
178388                          administración   
250439                                    <NA>   
26764                                     <NA>   
174483                            contabilidad   
209368  educación en computacion e informatica   
223922           fisioterapia y rehabilitacion   
292701                                    <NA>   
300534                                    <NA>   

                                                      p510_label  \
104908                                 empresa o patrono privado   
3892                                                        <NA>   
158463                                                      <NA>   
130769                                                      <NA>   
87795                                                       <NA>   
105899                                 empresa o patrono privado   
23688                                  empresa o patrono privado   
40554                                                       <NA>   
218262                                 empresa o patrono privado   
277073                                 empresa o patrono privado   
92733                                                       <NA>   
281676                                                      <NA>   
81081                                                       <NA>   
175664                                                      <NA>   
196868                                                      <NA>   
17525                                                       <NA>   
154354                                                      <NA>   
183663  fuerzas armadas, policía nacional del perú (militares)   
203010                                 empresa o patrono privado   
169875                                                      <NA>   
14569                                                  

In [50]:
df_sub["p510_label"] = df_sub["p510_label"].astype(str).apply(ftfy.fix_text)
display(df_sub[["p510_label"]].sample(5, random_state=42))

p510_label
104908  empresa o patrono privado
3892                         <NA>
158463                       <NA>
130769                       <NA>
87795                        <NA>

In [51]:
display(df_sub[["p301a1_label","p510_label", "p301a_label", "p507_label"]].sample(20, random_state=42))

p301a1_label  \
104908                                  <NA>   
3892                                    <NA>   
158463                                  <NA>   
130769                                  <NA>   
87795                                   <NA>   
105899                   mecanica automotriz   
23688                   ingeniería ambiental   
40554                                   <NA>   
218262                                  <NA>   
277073                                  <NA>   
92733                                   <NA>   
281676                                  <NA>   
81081                                   <NA>   
175664                                  <NA>   
196868                                  <NA>   
17525                                   <NA>   
154354                                  <NA>   
183663  administración y ciencias policiales   
203010              electrotecnia industrial   
169875                                  <NA>   

                                                    p510_label  \
104908                               empresa o patrono privado   
3892                                                      <NA>   
158463                                                    <NA>   
130769                                                    <NA>   
87795                                                     <NA>   
105899                               empresa o patrono privado   
23688                                empresa o patrono privado   
40554                                                     <NA>   
218262                               empresa o patrono privado   
277073                               empresa o patrono privado   
92733                                                     <NA>   
281676                                                    <NA>   
81081                                                     <NA>   
175664                                                    <NA>   
196868                                                    <NA>   
17525                                                     <NA>   
154354                                                    <NA>   
183663  fuerzas armadas, policía nacional del perú (militares)   
203010                               empresa o patrono privado   
169875                                                    <NA>   

                               p301a_label                         p507_label  
104908                 secundaria completa                           empleado  
3892                 secundaria incompleta  trabajador familiar no remunerado  
158463                 secundaria completa           trabajador independiente  
130769                   primaria completa  trabajador familiar no remunerado  
87795                  primaria incompleta  trabajador familiar no remunerado  
105899  superior no universitaria completa                             obrero  
23688    superior universitaria incompleta                               otro  
40554                  secundaria completa  trabajador familiar no remunerado  
218262                 secundaria completa                             obrero  
277073               secundaria incompleta                             obrero  
92733                  primaria incompleta           trabajador independiente  
281676                 secundaria completa               trabajador del hogar  
81081                  secundaria completa           trabajador independiente  
175664                 primaria incompleta           trabajador independiente  
196868                   primaria completa  trabajador familiar no remunerado  
17525                  primaria incompleta  trabajador familiar no remunerado  
154354                           sin nivel           trabajador independiente  
183663  superior no universitaria completa                           empleado  
203010  superior no universitaria completa                           empleado  
169875               secundaria inco

In [52]:
# ============================================================================
# CELDA 12: INSPECTOR DE CORRECCIONES
# ============================================================================
from difflib import get_close_matches

palabra_buscar = "cultibo"  # Cambia la palabra a inspeccionar

# Verificamos si el corrector tiene una sugerencia directa
correccion = spell.correction(palabra_buscar)
candidatos = spell.candidates(palabra_buscar)

print(f"🔍 Análisis para la palabra: '{palabra_buscar}'\n")

if correccion and correccion != palabra_buscar:
    print(f"✅ Sugerencia principal: '{correccion}'")
else:
    print(f"⚠️ No hay corrección segura registrada para '{palabra_buscar}'")

if candidatos:
    print("\n📋 Candidatos similares encontrados por SpellChecker:")
    for cand in sorted(candidatos, key=lambda w: zipf_frequency(w, "es"), reverse=True):
        print(f"  • {cand} (frecuencia: {zipf_frequency(cand, 'es'):.2f})")
else:
    print("❌ No se encontraron candidatos similares en el vocabulario del corrector.")

# Búsqueda adicional en el vocabulario español (wordfreq)
print("\n📚 Palabras del vocabulario español similares:")
vocab_matches = get_close_matches(palabra_buscar, list(spanish_vocab), n=10, cutoff=0.8)
for m in vocab_matches:
    print(f"  • {m} (frecuencia: {zipf_frequency(m, 'es'):.2f})")

if not vocab_matches:
    print("⚠️ No se encontraron coincidencias cercanas en el vocabulario general.")


🔍 Análisis para la palabra: 'cultibo'

✅ Sugerencia principal: 'cultivo'

📋 Candidatos similares encontrados por SpellChecker:
  • cultivo (frecuencia: 4.32)

📚 Palabras del vocabulario español similares:
  • cultivo (frecuencia: 4.32)
  • culto (frecuencia: 4.33)
  • cultivos (frecuencia: 4.12)


In [53]:
import os
print("📂 Ruta actual:", os.getcwd())

📂 Ruta actual: c:\Users\Hugo Jazyel\OneDrive\INEI\PEU-CD\Cursos especializados\Proyecto Integrador


In [54]:
df_sub.to_parquet("BASE_LIMPIA_VF.parquet", index=False)
print("✅ Archivo guardado como BASE_LIMPIA_VF.parquet")

✅ Archivo guardado como BASE_LIMPIA_VF.parquet
